community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -142,7 +143,7 @@ TransactionWriteState upgradeToSchemaWrites() throws InvalidTransactionTypeKerne
     private TransactionHooks.TransactionHooksState hooksState;
     private final KernelStatement currentStatement;
     private final StorageStatement storageStatement;
-    private CloseListener closeListener;
+    private final List<CloseListener> closeListeners = new ArrayList<>( 2 );
```

Comment:
When we discussed this change we've reached a conclusion that `CloseListener` is not the best foundation to build upon. Now this PR even pushes the listeners thing forward. If we do not take a step back now to rethink the design, we'll find ourselves in a situation when proposed refactoring is really hard to do because too many things rely on it.